core/src/upgrade/transfer.rs

```diff
@@ -29,7 +29,7 @@ use std::{error, fmt, io};
 ///
 /// > **Note**: Prepends a variable-length prefix indicate the length of the message. This is
 /// >           compatible with what `read_one` expects.
-pub async fn write_one(socket: &mut (impl AsyncWrite + Unpin), data: impl AsRef<[u8]>)
+pub async fn write_and_close(socket: &mut (impl AsyncWrite + Unpin), data: impl AsRef<[u8]>)
```

Comment:
```suggestion
pub async fn write_with_len_prefix_and_close(socket: &mut (impl AsyncWrite + Unpin), data: impl AsRef<[u8]>)
```

For the sake of consistency with the below `fn` `write_with_len_prefix`, wouldn't this naming be better?

In the same spirit, what do you think of:

```diff
- pub async fn read_one(socket: &mut (impl AsyncRead + Unpin), max_size: usize)
+ pub async fn read_with_len_prefix(socket: &mut (impl AsyncRead + Unpin), max_size: usize)
```

Reply:
So actually, I would want to remove the `with_len_prefix` from all functions because that piece of logic is integral to all of them and hence not really worth pointing out IMO.

The issue with that is that we would effectively change `write_one` for `write_with_len_prefix` assuming we also keep the rename to `write_one_and_close`.

This discussion makes me think though: Usually when it is hard to put multiple aspects into a single name, the composition of the function might not be ideal. Closing a substream is a single function call. What if we remove the functionality of closing entirely and only offer one function that just does the encoding?

Reply:
Possible names: `encode_one` and `decode_one`.

Reply:
I find `encode_*` misleading.

Renaming `write_with_len_prefix` to `write_one` would mean a breaking change without a change at the type level. I would prefer to not do that.

How about removing the current `write_one` and retain `write_with_len_prefix`. Or maybe rename to `write_length_prefixed`?

Reply:
> I find `encode_*` misleading.

Out of curiosity, can you expand on that?
What would be your expectation when you see a function signature of:

```rust
async fn encode_one(stream: &mut impl AsyncWrite, message: &impl AsRef<[u8]>);
```

And how is that expectation different from what the function will be doing?

My thinking was: Serializing to bytes and prefixing with the varint length _is_ a form of encoding the message. We could have the function return a Vec instead of trying to write to the stream directly but that is just unnecessary copying.


> Renaming `write_with_len_prefix` to `write_one` would mean a breaking change without a change at the type level. I would prefer to not do that.

Yep agreed. I tried to say that in my last comment but you expressed it much better 😅

> How about removing the current `write_one` and retain `write_with_len_prefix`. Or maybe rename to `write_length_prefixed`?

Happy with `write_length_prefixed`. Will adapt code accordingly!

Reply:
> And how is that expectation different from what the function will be doing?
>
> My thinking was: Serializing to bytes and prefixing with the varint length is a form of encoding the message. We could have the function return a Vec instead of trying to write to the stream directly but that is just unnecessary copying.

More of a gut feeling. _Encoding_ in my head changes the data itself. _Writing_ simply takes the bytes and shoves them down the wire. Neither seems to be a perfect fit here, though _writing_ in my eyes is closest, given that the length is only prefixed but the actual data is untouched.